File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.js
const fs = require("fs");
const input = fs.readFileSync('/dev/stdin').toString().split("\n");

let line = input.slice();
let answer = ""
for (let str of line) {
  if (str == ".") {
    break
  }

  let stack = [];

  for (let char of str) {
    if (char == "(" || char == "[") {
      stack.push(char);
    } else if (char == ")" || char == "]") {
      if (stack[stack.length - 1] == "(" && char == ")") {
        stack.pop()
      } else if (stack[stack.length - 1] == "[" && char == "]") {
        stack.pop()
      } else {
        stack.push(char)
      }
    }
  }

  answer += stack.length >= 1 ? "no\n" : "yes\n"
}

console.log(answer)

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
import sys

while True:
    line = sys.stdin.readline().rstrip()
    if line == ".":
        break
    
    stack = []
    is_valid = True
    
    for c in line:
        if c == "[" or c == "(":
            stack.append(c)
        elif c == "]" or c == ")":
            if not stack:  # 스택이 비어있으면
                is_valid = False
                break
            
            if stack[-1] == "[" and c == "]":
                stack.pop()
            elif stack[-1] == "(" and c == ")":
                stack.pop()
            else:  # 매칭되지 않으면
                is_valid = False
                break
    
    # 스택에 남은 괄호가 있거나 중간에 실패했으면 no
    if stack or not is_valid:
        print("no")
    else:
        print("yes")